File: app-modules-sample/commonLibrary/src/main/java/com/common/widget/status/StatusLayout.java
package com.common.widget.status;


import android.content.Context;
import android.content.res.TypedArray;
import android.support.annotation.AttrRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringDef;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.View;
import android.widget.FrameLayout;


import com.common.R;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * by y on 14/07/2017.
 * <p>
 * 多种状态布局Layout
 * <p>
 * 不一定要把所有的状态View都填充，例如有些页面没必要设置 ErrorView，可以不用填充
 * <p>
 * 建议 在 BaseActivity 或者 BaseFragment 里面  设置 StatusLayout 为 RootView，这样想改变页面状态很简单就完成了
 * <p>
 * <p>
 * 使用示例：
 * <pre>
 *     <StatusLayout
 *            android:id="@+id/statusLayout"
 *            android:layout_width="match_parent"
 *            android:layout_height="300dp"
 *            app:status_empty_layout="@layout/layout_empty"
 *            app:status_error_layout="@layout/layout_error"
 *            app:status_loading_layout="@layout/layout_loading"
 *            app:status_normal_layout="@layout/layout_normal"
 *            app:status_success_layout="@layout/layout_success" />
 *
 * or
 *
 *     statusLayout.setNorMalView(R.layout.layout_normal, null);
 *     statusLayout.setLoadingView(R.layout.layout_loading, null);
 *     statusLayout.setEmptyView(R.layout.layout_empty, null);
 *     statusLayout.setSuccessView(R.layout.layout_success, null);
 *     statusLayout.setErrorView(R.layout.layout_error, null);
 */

public class StatusLayout extends FrameLayout {

    private static final int NO_LAYOUT = 0x00;

    private String mStatus = null;
    private StatusClickListener clickListener = null;
    private View mNorMalView = null;
    private View mLoadingView = null;
    private View mEmptyView = null;
    private View mSuccessView = null;
    private View mErrorView = null;
    @LayoutRes
    private int mNormalLayoutId = NO_LAYOUT;
    @LayoutRes
    private int mLoadingLayoutId = NO_LAYOUT;
    @LayoutRes
    private int mEmptyLayoutId = NO_LAYOUT;
    @LayoutRes
    private int mSuccessLayoutId = NO_LAYOUT;
    @LayoutRes
    private int mErrorLayoutId = NO_LAYOUT;
    public StatusLayout(@NonNull Context context) {
        super(context);
        init(null);
    }

    public StatusLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(attrs);
    }

    public StatusLayout(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(attrs);
    }

    private void init(AttributeSet attrs) {
        TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.StatusLayout);

        mNormalLayoutId = typedArray.getResourceId(R.styleable.StatusLayout_status_normal_layout, NO_LAYOUT);
        mLoadingLayoutId = typedArray.getResourceId(R.styleable.StatusLayout_status_loading_layout, NO_LAYOUT);
        mEmptyLayoutId = typedArray.getResourceId(R.styleable.StatusLayout_status_empty_layout, NO_LAYOUT);
        mSuccessLayoutId = typedArray.getResourceId(R.styleable.StatusLayout_status_success_layout, NO_LAYOUT);
        mErrorLayoutId = typedArray.getResourceId(R.styleable.StatusLayout_status_error_layout, NO_LAYOUT);


        boolean normalFlag = typedArray.getBoolean(R.styleable.StatusLayout_status_normal_flag, false);
        boolean loadingFlag = typedArray.getBoolean(R.styleable.StatusLayout_status_loading_flag, false);
        boolean emptyFlag = typedArray.getBoolean(R.styleable.StatusLayout_status_empty_flag, false);
        boolean successFlag = typedArray.getBoolean(R.styleable.StatusLayout_status_success_flag, false);
        boolean errorFlag = typedArray.getBoolean(R.styleable.StatusLayout_status_error_flag, false);

        if (mNormalLayoutId != NO_LAYOUT) {
            if (normalFlag) {
                setNorMalView(mNormalLayoutId);
            } else {
                setNorMalView(mNormalLayoutId, null);
            }
        }
        if (mLoadingLayoutId != NO_LAYOUT) {
            if (loadingFlag) {
                setLoadingView(mLoadingLayoutId);
            } else {
                setLoadingView(mLoadingLayoutId, null);
            }
        }
        if (mEmptyLayoutId != NO_LAYOUT) {
            if (emptyFlag) {
                setEmptyView(mEmptyLayoutId);
            } else {
                setEmptyView(mEmptyLayoutId, null);
            }
        }
        if (mSuccessLayoutId != NO_LAYOUT) {
            if (successFlag) {
                setSuccessView(mSuccessLayoutId);
            } else {
                setSuccessView(mSuccessLayoutId, null);
            }
        }
        if (mErrorLayoutId != NO_LAYOUT) {
            if (errorFlag) {
                setErrorView(mErrorLayoutId);
            } else {
                setErrorView(mErrorLayoutId, null);
            }
        }

        typedArray.recycle();

        Util.goneView(mNorMalView, mLoadingView, mEmptyView, mSuccessView, mErrorView);
    }

    /**
     * 改变布局的状态
     *
     * @param status see{@link Status}
     * @return false : 当前状态和设置的状态一致.
     */
    public boolean setStatus(@Status String status) {
        if (TextUtils.equals(mStatus, status)) {
            return false;
        }
        Util.goneView(mNorMalView, mLoadingView, mEmptyView, mSuccessView, mErrorView);
        switch (status) {
            case Status.NORMAL:
                Util.visibilityView(mNorMalView);
                break;
            case Status.LOADING:
                Util.visibilityView(mLoadingView);
                break;
            case Status.EMPTY:
                Util.visibilityView(mEmptyView);
                break;
            case Status.SUCCESS:
                Util.visibilityView(mSuccessView);
                break;
            case Status.ERROR:
                Util.visibilityView(mErrorView);
                break;
            default:
                throw new RuntimeException("please check status");
        }
        mStatus = status;
        return true;
    }

    public void setStatusClickListener(StatusClickListener clickListener) {
        this.clickListener = clickListener;
    }

    /**********************   LayoutRes  默认 params ， 填充屏幕并居中   ************/

    public void setNorMalView(@LayoutRes int normalLayoutRes) {
        setNorMalView(Util.getViewLayout(this, normalLayoutRes));
    }

    public void setLoadingView(@LayoutRes int loadingLayoutRes) {
        setLoadingView(Util.getViewLayout(this, loadingLayoutRes));
    }

    public void setEmptyView(@LayoutRes int emptyLayoutRes) {
        setEmptyView(Util.getViewLayout(this, emptyLayoutRes));
    }

    public void setSuccessView(@LayoutRes int successLayoutRes) {
        setSuccessView(Util.getViewLayout(this, successLayoutRes));
    }

    public void setErrorView(@LayoutRes int errorLayoutRes) {
        setErrorView(Util.getViewLayout(this, errorLayoutRes));
    }

    /**********************   LayoutRes, params（可为 Null）  ************/

    public void setNorMalView(@LayoutRes int normalLayoutRes, @Nullable LayoutParams params) {
        setNorMalView(Util.getViewLayout(this, normalLayoutRes), params);
    }

    public void setLoadingView(@LayoutRes int loadingLayoutRes, @Nullable LayoutParams params) {
        setLoadingView(Util.getViewLayout(this, loadingLayoutRes), params);
    }

    public void setEmptyView(@LayoutRes int emptyLayoutRes, @Nullable LayoutParams params) {
        setEmptyView(Util.getViewLayout(this, emptyLayoutRes), params);
    }

    public void setSuccessView(@LayoutRes int successLayoutRes, @Nullable LayoutParams params) {
        setSuccessView(Util.getViewLayout(this, successLayoutRes), params);
    }

    public void setErrorView(@LayoutRes int errorLayoutRes, @Nullable LayoutParams params) {
        setErrorView(Util.getViewLayout(this, errorLayoutRes), params);
    }

    /**********************   最终填充View方法  flag addView() 的时候是否 使用 params   ************/

    public void setNorMalView(@NonNull View norMalView, @Nullable LayoutParams params) {
        if (mNorMalView != null) {
            removeView(mNorMalView);
            mNorMalView = null;
        }
        this.mNorMalView = norMalView;
        if (params != null) {
            addView(norMalView, params);
        } else {
            addView(norMalView);
        }
        mNorMalView.setOnClickListener(v -> {
            if (clickListener != null)
                clickListener.onNorMalClick();
        });
    }

    public void setLoadingView(@NonNull View loadingView, @Nullable LayoutParams params) {
        if (mLoadingView != null) {
            removeView(mLoadingView);
            mLoadingView = null;
        }
        this.mLoadingView = loadingView;
        if (params != null) {
            addView(loadingView, params);
        } else {
            addView(loadingView);
        }
        mLoadingView.setOnClickListener(v -> {
            if (clickListener != null)
                clickListener.onLoadingClick();
        });
    }

    public void setEmptyView(@NonNull View emptyView, @Nullable LayoutParams params) {
        if (mEmptyView != null) {
            removeView(mEmptyView);
            mEmptyView = null;
        }
        this.mEmptyView = emptyView;
        if (params != null) {
            addView(emptyView, params);
        } else {
            addView(emptyView);
        }
        mEmptyView.setOnClickListener(v -> {
            if (clickListener != null)
                clickListener.onEmptyClick();
        });
    }

    public void setSuccessView(@NonNull View successView, @Nullable LayoutParams params) {
        if (mSuccessView != null) {
            removeView(mSuccessView);
            mSuccessView = null;
        }
        this.mSuccessView = successView;
        if (params != null) {
            addView(successView, params);
        } else {
            addView(successView);
        }
        mSuccessView.setOnClickListener(v -> {
            if (clickListener != null)
                clickListener.onSuccessClick();
        });
    }

    public void setErrorView(@NonNull View errorView, @Nullable LayoutParams params) {
        if (mErrorView != null) {
            removeView(mErrorView);
            mErrorView = null;
        }
        this.mErrorView = errorView;
        if (params != null) {
            addView(errorView, params);
        } else {
            addView(errorView);
        }
        mErrorView.setOnClickListener(v -> {
            if (clickListener != null)
                clickListener.onErrorClick();
        });
    }

    public String getStatus() {
        return mStatus;
    }

    /************   返回View，有可能为 Null   ***********/

    @Nullable
    public View getNorMalView() {
        return mNorMalView;
    }

    /**********************   view  默认 params ， 填充屏幕   ************/


    public void setNorMalView(@NonNull View norMalView) {
        setNorMalView(norMalView, Util.getParams());
    }

    @Nullable
    public View getLoadingView() {
        return mLoadingView;
    }

    public void setLoadingView(@NonNull View loadingView) {
        setLoadingView(loadingView, Util.getParams());
    }

    @Nullable
    public View getEmptyView() {
        return mEmptyView;
    }

    public void setEmptyView(@NonNull View emptyView) {
        setEmptyView(emptyView, Util.getParams());
    }

    @Nullable
    public View getSuccessView() {
        return mSuccessView;
    }

    public void setSuccessView(@NonNull View successView) {
        setSuccessView(successView, Util.getParams());
    }

    @Nullable
    public View getErrorView() {
        return mErrorView;
    }

    public void setErrorView(@NonNull View errorView) {
        setErrorView(errorView, Util.getParams());
    }

    @StringDef({
            Status.NORMAL,
            Status.LOADING,
            Status.EMPTY,
            Status.SUCCESS,
            Status.ERROR
    })
    @Retention(RetentionPolicy.SOURCE)

    public @interface Status {
        String NORMAL = "StatusLayout:Normal"; // 初始状态
        String LOADING = "StatusLayout:Loading"; // 正在加载
        String EMPTY = "StatusLayout:Empty"; // 加载空布局
        String SUCCESS = "StatusLayout:Success"; // 加载成功
        String ERROR = "StatusLayout:Error"; // 加载失败
    }

}

File: StatusLayout/StatusLibrary/src/main/java/com/status/layout/StatusLayout.kt
@file:Suppress("MemberVisibilityCanBePrivate")

package com.status.layout

import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import androidx.annotation.LayoutRes
import androidx.annotation.StringDef

val params: FrameLayout.LayoutParams = FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT, Gravity.CENTER)

fun getViewLayout(statusLayout: StatusLayout, @LayoutRes id: Int): View = LayoutInflater.from(statusLayout.context).inflate(id, statusLayout, false)

fun goneView(vararg views: View?) {
    for (view in views) {
        if (view != null && view.visibility != View.GONE) {
            view.visibility = View.GONE
        }
    }
}

fun visibilityView(view: View?) {
    if (view != null && view.visibility != View.VISIBLE) {
        view.visibility = View.VISIBLE
    }
}

@StringDef(StatusLayout.NORMAL, StatusLayout.LOADING, StatusLayout.EMPTY, StatusLayout.SUCCESS, StatusLayout.ERROR)
@Retention(AnnotationRetention.SOURCE)
annotation class StatusAnnotation

interface OnStatusClickListener {
    fun onNorMalClick(view: View)
    fun onLoadingClick(view: View)
    fun onEmptyClick(view: View)
    fun onSuccessClick(view: View)
    fun onErrorClick(view: View)
}

open class SimpleOnStatusClickListener : OnStatusClickListener {
    override fun onNorMalClick(view: View) {}
    override fun onLoadingClick(view: View) {}
    override fun onEmptyClick(view: View) {}
    override fun onSuccessClick(view: View) {}
    override fun onErrorClick(view: View) {}
}

class StatusLayout @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : FrameLayout(context, attrs, defStyleAttr) {

    companion object {
        const val NORMAL = "StatusLayout:Normal"
        const val LOADING = "StatusLayout:Loading"
        const val EMPTY = "StatusLayout:Empty"
        const val SUCCESS = "StatusLayout:Success"
        const val ERROR = "StatusLayout:Error"
    }

    private var mStatus: String = NORMAL
    var onStatusClickListener: OnStatusClickListener? = null
    private var mNorMalView: View? = null
    private var mLoadingView: View? = null
    private var mEmptyView: View? = null
    private var mSuccessView: View? = null
    private var mErrorView: View? = null

    init {
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.StatusLayout)
        val mNormalLayoutId = typedArray.getResourceId(R.styleable.StatusLayout_status_normal_layout, View.NO_ID)
        val mLoadingLayoutId = typedArray.getResourceId(R.styleable.StatusLayout_status_loading_layout, View.NO_ID)
        val mEmptyLayoutId = typedArray.getResourceId(R.styleable.StatusLayout_status_empty_layout, View.NO_ID)
        val mSuccessLayoutId = typedArray.getResourceId(R.styleable.StatusLayout_status_success_layout, View.NO_ID)
        val mErrorLayoutId = typedArray.getResourceId(R.styleable.StatusLayout_status_error_layout, View.NO_ID)
        typedArray.recycle()
        if (mNormalLayoutId != View.NO_ID) {
            addNorMalView(mNormalLayoutId)
        }
        if (mLoadingLayoutId != View.NO_ID) {
            addLoadingView(mLoadingLayoutId)
        }
        if (mEmptyLayoutId != View.NO_ID) {
            addEmptyView(mEmptyLayoutId)
        }
        if (mSuccessLayoutId != View.NO_ID) {
            addSuccessView(mSuccessLayoutId)
        }
        if (mErrorLayoutId != View.NO_ID) {
            addErrorView(mErrorLayoutId)
        }
        goneView(mNorMalView, mEmptyView, mErrorView, mLoadingView, mSuccessView)
    }

    fun setStatus(@StatusAnnotation status: String): Boolean {
        if (TextUtils.equals(mStatus, status)) {
            return false
        }
        when (mStatus) {
            NORMAL -> goneView(mNorMalView)
            LOADING -> goneView(mLoadingView)
            EMPTY -> goneView(mEmptyView)
            SUCCESS -> goneView(mSuccessView)
            ERROR -> goneView(mErrorView)
            else -> throw RuntimeException("please check status")
        }
        when (status) {
            NORMAL -> visibilityView(mNorMalView)
            LOADING -> visibilityView(mLoadingView)
            EMPTY -> visibilityView(mEmptyView)
            SUCCESS -> visibilityView(mSuccessView)
            ERROR -> visibilityView(mErrorView)
            else -> throw RuntimeException("please check status")
        }
        mStatus = status
        return true
    }

    fun addNorMalView(norMalView: View): StatusLayout {
        addNorMalView(norMalView, params)
        return this
    }

    fun addLoadingView(loadingView: View): StatusLayout {
        addLoadingView(loadingView, params)
        return this
    }

    fun addEmptyView(emptyView: View): StatusLayout {
        addEmptyView(emptyView, params)
        return this
    }

    fun addSuccessView(successView: View): StatusLayout {
        addSuccessView(successView, params)
        return this
    }

    fun addErrorView(errorView: View): StatusLayout {
        addErrorView(errorView, params)
        return this
    }

    fun addNorMalView(@LayoutRes normalLayoutRes: Int): StatusLayout {
        addNorMalView(getViewLayout(this, normalLayoutRes))
        return this
    }

    fun addLoadingView(@LayoutRes loadingLayoutRes: Int): StatusLayout {
        addLoadingView(getViewLayout(this, loadingLayoutRes))
        return this
    }

    fun addEmptyView(@LayoutRes emptyLayoutRes: Int): StatusLayout {
        addEmptyView(getViewLayout(this, emptyLayoutRes))
        return this
    }

    fun addSuccessView(@LayoutRes successLayoutRes: Int): StatusLayout {
        addSuccessView(getViewLayout(this, successLayoutRes))
        return this
    }

    fun addErrorView(@LayoutRes errorLayoutRes: Int): StatusLayout {
        addErrorView(getViewLayout(this, errorLayoutRes))
        return this
    }

    fun addNorMalView(@LayoutRes normalLayoutRes: Int, params: FrameLayout.LayoutParams?): StatusLayout {
        addNorMalView(getViewLayout(this, normalLayoutRes), params)
        return this
    }

    fun addLoadingView(@LayoutRes loadingLayoutRes: Int, params: FrameLayout.LayoutParams?): StatusLayout {
        addLoadingView(getViewLayout(this, loadingLayoutRes), params)
        return this
    }

    fun addEmptyView(@LayoutRes emptyLayoutRes: Int, params: FrameLayout.LayoutParams?): StatusLayout {
        addEmptyView(getViewLayout(this, emptyLayoutRes), params)
        return this
    }

    fun addSuccessView(@LayoutRes successLayoutRes: Int, params: FrameLayout.LayoutParams?): StatusLayout {
        addSuccessView(getViewLayout(this, successLayoutRes), params)
        return this
    }

    fun addErrorView(@LayoutRes errorLayoutRes: Int, params: FrameLayout.LayoutParams?): StatusLayout {
        addErrorView(getViewLayout(this, errorLayoutRes), params)
        return this
    }

    fun addNorMalView(norMalView: View, params: FrameLayout.LayoutParams?): StatusLayout {
        if (mNorMalView != null) {
            removeView(mNorMalView)
            mNorMalView = null
        }
        this.mNorMalView = norMalView
        if (params != null) {
            addView(norMalView, params)
        } else {
            addView(norMalView)
        }
        mNorMalView?.setOnClickListener { v -> onStatusClickListener?.onNorMalClick(v) }
        return this
    }

    fun addLoadingView(loadingView: View, params: FrameLayout.LayoutParams?): StatusLayout {
        if (mLoadingView != null) {
            removeView(mLoadingView)
            mLoadingView = null
        }
        this.mLoadingView = loadingView
        if (params != null) {
            addView(loadingView, params)
        } else {
            addView(loadingView)
        }
        mLoadingView?.setOnClickListener { v -> onStatusClickListener?.onLoadingClick(v) }
        return this
    }

    fun addEmptyView(emptyView: View, params: FrameLayout.LayoutParams?): StatusLayout {
        if (mEmptyView != null) {
            removeView(mEmptyView)
            mEmptyView = null
        }
        this.mEmptyView = emptyView
        if (params != null) {
            addView(emptyView, params)
        } else {
            addView(emptyView)
        }
        mEmptyView?.setOnClickListener { v -> onStatusClickListener?.onEmptyClick(v) }
        return this
    }

    fun addSuccessView(successView: View, params: FrameLayout.LayoutParams?): StatusLayout {
        if (mSuccessView != null) {
            removeView(mSuccessView)
            mSuccessView = null
        }
        this.mSuccessView = successView
        if (params != null) {
            addView(successView, params)
        } else {
            addView(successView)
        }
        mSuccessView?.setOnClickListener { v -> onStatusClickListener?.onSuccessClick(v) }
        return this
    }

    fun addErrorView(errorView: View, params: FrameLayout.LayoutParams?): StatusLayout {
        if (mErrorView != null) {
            removeView(mErrorView)
            mErrorView = null
        }
        this.mErrorView = errorView
        if (params != null) {
            addView(errorView, params)
        } else {
            addView(errorView)
        }
        mErrorView?.setOnClickListener { v -> onStatusClickListener?.onErrorClick(v) }
        return this
    }

    @StatusAnnotation
    fun getStatus(): String {
        return mStatus
    }

    fun getView(@StatusAnnotation status: String): View? {
        return when (status) {
            NORMAL -> mNorMalView
            LOADING -> mLoadingView
            EMPTY -> mEmptyView
            SUCCESS -> mSuccessView
            ERROR -> mErrorView
            else -> throw RuntimeException("please check status")
        }
    }
}

